### DISCUSSION TITLE
Using Realtime fetch

### DISCUSSION THREAD

**Discussion by shamxeed** [TanStack/query#2025] [Q&A]
How is real-time fetching is going to affect my server performance? Is it recommended?

**Comment by TkDodo** [comment]
What is real-time fetching please?

**Reply by shamxeed** [reply]
An example [here](https://github.com/tannerlinsley/react-query/blob/69d88cca55915a1830d66465289ecc2a6e83329f/examples/auto-refetching/pages/index.js), shows that we can fetch data after every second using refetchInterval. How is that going to affect server performance and is it even recommended? Thanks!!

**Comment by TkDodo** [comment]
okay, interval fetching :)

that is up to you to decide. do you need data fetched every N seconds? N minutes? is it needed at all? depends on the use-case. I can't tell you if your server can handle it

**Reply by shamxeed** [reply]
Indeed I need that but I'm afraid my server can not handle it especially when the requests become massive as a result of a lot of users. Perhaps I can try optimistic updates then.

**Reply by TkDodo** [reply]
can you elaborate on your use-case? which data are you presenting, and how would optimistic updates make a difference here?

**Reply by shamxeed** [reply]
It is a chatting App that I want an update to happen after every sendChat mutation. Or maybe I can also invalidateQuery that needs to be updated after every sendChat mutaion?

**Reply by TkDodo** [reply]
For sending, I would do a mutation with optimistic update and invalidate onSettled. Like the optimistic update example in the docs.

for real-time fetching, a websocket would probably be better to push info. See e.g. https://github.com/tannerlinsley/react-query/issues/171#issuecomment-649716718

**Reply by shamxeed** [reply]
Okay! But in the optimistic update example, the use case is for an array of strings I guess, while in my use case I'm dealing with an array of objects. If you can guide me with some workarounds I will appreciate that.

**Reply by TkDodo** [reply]
why does that matter? Optimistic updates are just:

- you do a mutation with variables of type X
- in the onMutate, you grab the current value from the cache, whatever that may be.
- you optimistically write the data in the cache. If your cache value is an array of objects, merge your variables into it at the correct place. it depends on your use-case
- return the old value from onMutate so that you can make a rollback onError
- invalidate in onSettled
- profit

**Reply by shamxeed** [reply]
Okay, thanks a lot for your time and commitment, I did it. Mind seeing how I achieve that? maybe for future references.

**Reply by TkDodo** [reply]
🥳
Sure, please share 👍

**Reply by shamxeed** [reply]
Showing just how I updated the cached query.   

```js
 queryClient.setQueryData('myInbox', old => {
      const toBeChanged = old.myInbox.find(i => i.id === inboxId)
      toBeChanged.chat = [...toBeChanged.chat, newData]
      return { ...old }
  })
```
BOOM!!! the cached query is been updated with the new data I passed to it. Thanks once again dude!!!!
